internal/debug/debug.go

```diff
@@ -0,0 +1,53 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug // import "go.opentelemetry.io/otel/internal/debug"
+
+import (
+	"log"
+	"os"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/stdr"
+)
+
+// globalLogger is the logging interface used within the otel api and sdk provide deatails of the internals.
+//
+// The default logger uses stdr which is backed by the standard `log.Logger`
+// interface. This logger will only show messages at the Error Level.
+var globalLoggger logr.Logger = stdr.New(log.New(os.Stderr, "", log.LstdFlags|log.Lshortfile))
+
+// SetLogger overrides the globalLogger with l.
+//
+// To see Info messages use a logger with `l.V(1).Enabled() == true`
+// To see Debug messages use a logger with `l.V(5).Enabled() == true`
+func SetLogger(l logr.Logger) {
+	globalLoggger = l
```

Comment:
Because I removed the exported value this does have those guarantees.  Any call made after a SetLogger will be routed to the new logger, and because there is no way to capture this logger you don't need lazy evaluation or delegation.

Reply:
Gotcha :+1:.

There does still seem to be a race condition here where one of the logging functions is called concurrently with `SetLogger`.

Reply:
Can you explain more?  Because there is no check and then use we can't race here.  Yes if you call a log function before you call Set it will go to the old one, but there is no locking that would fix that.

Reply:
Maybe I'm missing something, but one goroutine calls `SetLogger` and another goroutine calls, say, `Info` since one is assigning to `globalLogger` and the other is reading it there is a race condition.

For example:

```
package debug

import (
	"log"
	"os"
	"testing"

	"github.com/go-logr/stdr"
)

func TestRace(t *testing.T) {
	go SetLogger(stdr.New(log.New(os.Stderr, "", 0)))
	go Info("")
}
```

and

```
go test -race
PASS
==================
WARNING: DATA RACE
Read at 0x0000006c2c50 by goroutine 9:
  runtime.racereadrange()
      <autogenerated>:1 +0x1b
  go.opentelemetry.io/otel/internal/debug.TestRace·dwrap·2()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:13 +0x71

Previous write at 0x0000006c2c50 by goroutine 8:
  go.opentelemetry.io/otel/internal/debug.SetLogger()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug.go:36 +0x4b
  go.opentelemetry.io/otel/internal/debug.TestRace·dwrap·1()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:12 +0x58

Goroutine 9 (running) created at:
  go.opentelemetry.io/otel/internal/debug.TestRace()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:13 +0x26e
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/lib/go/src/testing/testing.go:1306 +0x47

Goroutine 8 (finished) created at:
  go.opentelemetry.io/otel/internal/debug.TestRace()
      /home/tyler/go/src/go.opentelemetry.io/otel/internal/debug/debug_test.go:12 +0x1cb
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/lib/go/src/testing/testing.go:1306 +0x47
==================
Found 1 data race(s)
exit status 66
FAIL	go.opentelemetry.io/otel/internal/debug	0.021s
```